File: backend/EntpFlow/Services/TaskCommentService.cs
using EntpFlow.Data;
using EntpFlow.Models;
using Microsoft.EntityFrameworkCore;

namespace EntpFlow.Services;

public class TaskCommentService
{
    private readonly ApplicationDbContext _context;

    public TaskCommentService(ApplicationDbContext context)
    {
        _context = context;
    }

    public async Task<IEnumerable<TaskComment>> GetComments(int taskId)
    {
        return await _context.TaskComments
            .Where(x => x.TaskId == taskId)
            .Include(x => x.User)
            .ToListAsync();
    }

    public async Task AddComment(TaskComment comment)
    {
        _context.TaskComments.Add(comment);
        await _context.SaveChangesAsync();
    }

    public async Task UpdateComment(int id, string newComment, int userId)
    {
        var comment = await _context.TaskComments.FindAsync(id);
        if (comment == null || comment.UserId != userId)
            throw new Exception("Comment not found or unauthorized");

        comment.Comment = newComment;
        await _context.SaveChangesAsync();
    }

    public async Task DeleteComment(int id, int userId)
    {
        var comment = await _context.TaskComments.FindAsync(id);
        if (comment == null || comment.UserId != userId)
            throw new Exception("Comment not found or unauthorized");

        _context.TaskComments.Remove(comment);
        await _context.SaveChangesAsync();
    }
}

File: Frontend/entpflow-ui/src/app/core/services/task-comment.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, tap } from 'rxjs';
import { TaskComment, CreateComment, UpdateComment } from '../models/task-comment.model';

@Injectable({
  providedIn: 'root',
})
export class TaskCommentService {

  private readonly api = 'https://localhost:7033/api/TaskComments';

  constructor(private http: HttpClient) {}

  getComments(taskId: number): Observable<TaskComment[]> {
    return this.http.get<TaskComment[]>(`${this.api}/${taskId}`);
  }

  addComment(payload: CreateComment): Observable<any> {
    return this.http.post(this.api, payload);
  }

  updateComment(id: number, payload: UpdateComment): Observable<any> {
    return this.http.put(`${this.api}/${id}`, payload);
  }

  deleteComment(id: number, userId: number): Observable<any> {
    return this.http.delete(`${this.api}/${id}`, { params: { userId: userId.toString() } });
  }
}
